api/books/book.service.ts

```diff
@@ -0,0 +1,84 @@
+import {Logger} from "../shared/logger.service";
+import {DEFAULT_REQUEST_DETAILS} from "../../koa.config";
+import {Book} from "./book.model";
+import {ApiError} from "../errors/api-error.errors";
+import {errors} from "../errors/errors";
+
+class BookService {
+
+    async create(bookData, details = DEFAULT_REQUEST_DETAILS) {
+        let book = new Book(bookData);
+        let savedBook = await book.save();
+        Logger.log("info", "[BookService] [save]");
+        let bookJson = savedBook.toJSON();
+        Logger.log("info", "[BookService] [Create] book created successfully", {book: bookJson, details});
+        return bookJson;
+    }
+
+    async findByIsbn(isbn, details = DEFAULT_REQUEST_DETAILS) {
+        let book = await Book.findOne({isbn: isbn});
+        if (!book) {
+            return null;
+        }
+
+        if (book && book.deletedAt) {
+            throw new ApiError(errors.generic.not_found);
+        }
+
+        Logger.log("info", "[BookService] [Find By Isbn] book retrieved successfully", {isbn, details});
+        return book.toJSON();
+    }
+
+    async search(query: any = {}, details = DEFAULT_REQUEST_DETAILS) {
+        let search: any = {};
+
+        if (query.q) {
+            search.$or = [
+                {"title": {$regex: query.q, $options: "i"}},
+                {"isbn": {$regex: query.q, $options: "i"}},
+            ];
+        }
+        if (query.active === undefined) {
+            query.active = "true";
+        }
+
+        if (query.active !== "both") {
+            search.deletedAt = {$exists: query.active !== "true"};
+        }
+
+        if (query.active !== "true") {
+            throw new ApiError(errors.generic.unauthorized);
+        }
+
+        let book = await Book.find(search).lean();
+
+        return book;
+    }
```

Comment:
how can we reduce the "if" filaters, maybe adding more parameters? seems like `query.active` has many responsabilities, this is minor is open to discussion

Reply:
Another alternative could be to define the _query_ type taking advantage of the [rich and very expressive TypeScript type system](https://www.typescriptlang.org/docs/handbook/advanced-types.html). So, it is possible declare what possible values could _active_ field take using _string literal types_ and maybe _tagged union types_ to identify different object structures based on _active_ value.

Anyway, as Mariano said, _active_ looks as a boolean field, maybe it is not the best name.